### PR TITLE
fix(swap): read the native balance when confirming a native sell

### PR DESCRIPTION
**Pull request by elena-zh**
## Summary

Regression from #8128: selling a **native** token shows *"Insufficient \<native\> balance"* in the Review Swap modal whenever the user's wrapped balance is lower than their native balance, even with plenty of native funds.
Reported in https://nomevlabs.slack.com/archives/C0361CDG8GP/p1789463250189059

image

**To test:** 
1. Connect to a wallet
2. make sure that you have a wrapped token balance < than a native token 
3. pick a native token in the sell field, > than a wrapped token amount in the wallet
4. proceed with the order placement
5. make sure that you don't see 'Insufficient balance' error in the confirm modal:
image
image

6. do a regression test and make sure that #8128 PR changes still work

<details>
## Root cause

#8128 correctly started checking the balance against the slippage-inclusive maximum sell amount, but it also moved the **currency used for the balance lookup** onto that amount:

```ts
const maximumSellAmount = receiveAmountInfo?.afterSlippage.sellAmount ?? inputCurrencyInfo?.amount
const current = maximumSellAmount?.currency        // was: inputCurrencyInfo?.amount?.currency
const balance = balances[getAddressKey(getCurrencyAddress(current))]
```

That currency comes from `useQuoteCurrencies`, which resolves it from `quoteResponse.quote.sellToken` (`useGetSwapReceiveAmountInfo.ts:62-64`). A CoW order can't use the native token, so for a native sell the quote is denominated in the **wrapped** token.

`getCurrencyAddress` then returns the wrapped contract address instead of the native sentinel (`libs/common-utils/src/getCurrencyAddress.ts:7-11`), so the check reads the user's **WXDAI** balance rather than their **xDAI** balance.

Reported case: selling 0.5 xDAI on Gnosis with 0 WXDAI held — needs 0.5025 WXDAI, has 0 → confirm blocked.

## Changes

Split the two roles. The currency the user actually spends picks the balance bucket; the maximum sell amount stays the figure that has to be covered. Extracted into `getIsBalanceEnough` in the existing `SwapConfirmModal.utils.ts` so it can be unit tested (the surrounding `useMemo` already carries a `TODO: Reduce function complexity by extracting logic`). The helper takes the input amount rather than a pre-picked currency, so the currency choice is covered by tests instead of sitting at the call site where #8128 broke it.

#8128's actual goal is unchanged — the balance still has to cover slippage.

**Note:** the comparison is between a wrapped-denominated amount and a native-denominated balance, so it is normalised by each side's `decimalScale` rather than comparing raw values. That matters because `CurrencyAmount.fromRawAmount` leaves the `Fraction` denominator at `1` (`currencyAmount.ts:19-41`), so `lessThan`/`equalTo` compare raw numerators — correct only while the two currencies share decimals. They do on every supported chain today, but the helper no longer depends on it.

The main trade form is unaffected: `validateTradeForm.ts` compares against `inputCurrencyBalance`, which is already the native balance, without re-looking it up by currency.

## Verification

- 9 new unit tests on `getIsBalanceEnough`, using real `NATIVE_CURRENCIES`/`WRAPPED_NATIVE_CURRENCIES` and `CurrencyAmount` rather than mocks, so the native/wrapped split is genuinely exercised.
- **Confirmed these catch the regression:** reintroducing the old lookup fails 4 of them, including `reads the native balance when selling the native token`; reverting to a raw (non-normalised) comparison fails `compares by value when the amount currency uses different decimals`.
- Full file suite: 19/19 passing. `tsc --noEmit` and eslint clean.
- Verified by code tracing and unit tests, **not yet in a browser** — the live check is selling a small amount of native xDAI on Gnosis with little or no WXDAI and confirming the Review Swap button is enabled.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved balance validation in the swap confirmation flow.
  - Confirmation is disabled when the available balance is insufficient or balance information is unavailable, without incorrectly labeling unavailable balances as insufficient.
  - Improved handling of native and wrapped-token balance comparisons, including exact matches.
  - Improved validation for slippage-adjusted maximum sell amounts and differing token decimal formats.
  - Prevented unrelated quoted currencies from being treated as insufficient balances.

### CODE REVIEW

**Review by elena-zh — COMMENTED** [review]
<details>
<summary>⚠️ <strong>AI Review</strong> (Claude Opus 5, worked 3m): the call site that caused the bug is untested; cross-currency comparison relies on equal decimals</summary>

**Finding:** [NON-BLOCKING] The regression this PR fixes would not be caught by the new tests

- Location: `apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/index.tsx:125-128`
- The defect introduced by #8128 was at the **call site** — passing `maximumSellAmount.currency` instead of the input currency. `getIsBalanceEnough` itself was never wrong.
- All 6 new tests exercise the helper directly. Changing line 125 back to `const sellCurrency = maximumSellAmount?.currency` leaves **16/16 green**, so the exact regression can be reintroduced without any test failing.
- `SwapConfirmModal` has no rendering test, so nothing covers the wiring.

**Suggested fix**
- Add a test at the modal boundary asserting the balance bucket is chosen from `inputCurrencyInfo`, not from the quote-derived amount — a focused render test with a native input and a zero wrapped balance, or extract the `maximumSellAmount`/`sellCurrency` pairing into a tested helper that the `useMemo` just calls.

**Finding:** [NON-BLOCKING] Cross-currency comparison silently depends on native and wrapped sharing decimals

- Location: `SwapConfirmModal.utils.ts:28-35`
- `maximumSellAmount` is denominated in the **wrapped** token while `balanceAsCurrencyAmount` is built from the **native** currency, and the two are compared directly.
- This is a **raw-value** comparison, not a decimals-normalized one: `CurrencyAmount.fromRawAmount` passes no denominator to `Fraction` (`libs/currency/src/entities/fractions/currencyAmount.ts:19-41`), so `decimalScale` is a separate field and `lessThan`/`equalTo` compare raw numerators (`fraction.ts:66-74`).
- I verified every supported chain currently matches (1/56/100/137/8453/9745/42161/43114/57073/59144/11155111 all 18/18, Solana 9/9), so this is correct today — but it is load-bearing, and the comment explaining it was removed.
- If a future chain ever ships a wrapped token with different decimals, the confirm gate silently mis-compares by orders of magnitude rather than failing loudly.

**Suggested fix**
- Compare raw values explicitly (`balance ?? 0n` against `maximumSellAmount.quotient`) so no `CurrencyAmount` of a mismatched currency is constructed, or add a decimals equality guard.

**Finding:** [QUESTION] Confirm gate not exercised in a browser

- Location: `index.tsx:123-140`
- This code decides whether a user can confirm a swap. The change is covered by unit tests and the PR body discloses that it was not browser-verified, so this is a traceability question rather than a blocker — a false result here blocks an order rather than misrouting one.
- Could you add the observed ER-vs-AR for a native sell on Gnosis with zero/low WXDAI (button enabled, order confirms), plus one wrapped-token sell to confirm no regression the other way?

<details>
<summary>Review scope and related context</summary>

- No existing review threads or reviews on this PR, so nothing here duplicates prior feedback. CodeRabbit skipped it while it was still a draft — it may re-review now that it is marked ready.
- Automated checks run locally on `d0722778b`: 16/16 in `SwapConfirmModal.utils.test.ts`, `tsc --noEmit` clean, ESLint clean apart from a pre-existing `import/no-internal-modules` warning that is already on `develop`. CI was still in progress at review time.
- Confirmed the PR body's claim that `validateTradeForm` is unaffected: `inputCurrencyBalance` comes from `useCurrencyAmountBalanceCombined(inputCurrency)` (`useBuildTradeDerivedState.ts:59`), i.e. already the native balance, with no re-lookup by currency.
- Checked and not raised: the cross-chain path is safe either way, since `getCrossChainReceiveAmountInfo` overrides `inputCurrency` from derived state; and `maximumSellAmount` can never be null while `sellCurrency` is set, so the `hasCurrentCurrency` / `isBalanceEnough` pairing matches the pre-#8128 behavior.
- Not raised as a finding: the confirm modal and `validateTradeForm` still compute balance sufficiency independently from two different sources, which is the structural reason this class of mismatch is possible. Pre-existing, worth a follow-up rather than this PR.

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

```text
Verify these findings against current code. Fix only if still valid, keep the change minimal, and validate with the targeted tests.

Context (apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/):

1. index.tsx:125 picks the balance bucket via `inputCurrencyInfo?.amount?.currency`. That line is the one #8128 broke,
   and no test covers it: reverting it to `maximumSellAmount?.currency` keeps all 16 tests in
   SwapConfirmModal.utils.test.ts passing. Add boundary coverage so the call site is guarded, not just the helper.

2. SwapConfirmModal.utils.ts:28-35 compares a wrapped-denominated `maximumSellAmount` against a native-denominated
   `balanceAsCurrencyAmount`. CurrencyAmount.fromRawAmount leaves the Fraction denominator at 1, so this is a raw
   comparison that is only correct while native and wrapped decimals match (they do on all chains today).
   Prefer comparing `balance ?? 0n` to `maximumSellAmount.quotient`, or guard on decimals equality.

Validate with:
pnpm exec nx test cowswap-frontend --testPathPatterns=SwapConfirmModal.utils --runInBand
```

</details>

_Generated using the `pr-review` skill from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

</details>

**Comment by elena-zh** [comment]
<details>
<summary>✅ <strong>AI Review</strong> (Claude Opus 5): both non-blocking findings addressed in f2abeecd5; the QA question is still open</summary>

**Rechecked**
- Code path: `SwapConfirmModal.utils.ts:28-38`, `index.tsx:125-131`.
- Test coverage: `pnpm exec nx test cowswap-frontend --testPathPatterns=SwapConfirmModal.utils --runInBand` — 19/19.

**Finding 1 (untested call site): fixed.** `getIsBalanceEnough` now takes `inputAmount` instead of a pre-picked `sellCurrency`, so both the currency choice and the `?? inputAmount` fallback happen inside the helper. The call site is pure data passing.

Reintroducing the original #8128 bug inside the helper (`const sellCurrency = (maximumSellAmount ?? inputAmount)?.currency`) now fails **4 of 19** tests. Before this commit the same mistake at the call site failed none.

**Finding 2 (raw cross-currency comparison): fixed.** The comparison is now normalised by each side's scale:

```ts
const sellCurrencyScale = 10n ** BigInt(sellCurrency.decimals)

return amountToCover.quotient * sellCurrencyScale <= balance * amountToCover.decimalScale
```

No `CurrencyAmount` of a mismatched currency is constructed, and the equal-decimals assumption is gone rather than merely documented. Reverting to `amountToCover.quotient <= balance` fails the new `compares by value when the amount currency uses different decimals` test.

**Finding 3 (browser verification): still open.** Unchanged — this needs a real browser and I have not run one. The ask stands: ER-vs-AR for a native sell on Gnosis with zero/low WXDAI, plus one wrapped-token sell as a control.

Test count is 9 on the helper (was 6); the three additions cover mismatched decimals, the slippage-maximum vs raw-input distinction, and the no-maximum fallback. `tsc --noEmit` and eslint are clean apart from the pre-existing `import/no-internal-modules` warning already on `develop`. The PR body has been corrected — the earlier note argued the comparison was safe *because* native and wrapped share decimals, which is no longer the reason it holds.

<details>
<summary>🤖 Verification notes for AI agents</summary>

```text
Verify the prior findings against current code only. Do not reopen broader or unrelated issues.

Prior findings:
1. index.tsx picked the balance bucket at the call site, untested; reverting it kept all tests green.
2. getIsBalanceEnough compared a wrapped-denominated amount against a native-denominated balance
   using raw Fraction numerators, correct only while decimals match.

Current verification:
- getIsBalanceEnough now accepts inputAmount and derives sellCurrency/amountToCover internally.
- Comparison normalises both sides: amountToCover.quotient * 10^sellCurrency.decimals
  <= balance * amountToCover.decimalScale.
- pnpm exec nx test cowswap-frontend --testPathPatterns=SwapConfirmModal.utils --runInBand => 19/19.
- Mutation checks: bug 1 reintroduced => 4 failures; bug 2 reintroduced => 1 failure.
```

</details>

*Generated using the `pr-review` skill from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills).*

</details>

File: apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/index.tsx [hunk 1]
```diff
@@ -125,24 +122,13 @@ export function SwapConfirmModal(props: SwapConfirmModalProps): ReactNode {
...
-
-      isBalanceEnough = Boolean(
-        balanceAsCurrencyAmount &&
-          maximumSellAmount &&
-          (maximumSellAmount.equalTo(balanceAsCurrencyAmount) || maximumSellAmount.lessThan(balanceAsCurrencyAmount)),
-      )
-    }
+    const inputAmount = inputCurrencyInfo?.amount
+    const hasCurrentCurrency = Boolean(inputAmount?.currency)
+    const isBalanceEnough = getIsBalanceEnough({
```
Comment (limitofzero) [inline apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/index.tsx:124]: nitpick: we should be quite explicit here since we compare different currencies, so I would suggest to add the additional check here:

  ```ts
  export function getIsBalanceEnough({ inputAmount, maximumSellAmount, balances, areBalancesReady }): boolean | null {
    const sellCurrency = inputAmount?.currency
    const amountToCover = maximumSellAmount ?? inputAmount

    if (!sellCurrency || !amountToCover) return null
    if (!areBalancesReady) return null
    // explicit check, there is only one case when currency could be different - when the sell token is native one
    if (!isQuotedInSameAsset(sellCurrency, amountToCover.currency)) return null
    …
  }

  function isQuotedInSameAsset(sellCurrency: Currency, quotedCurrency: Currency): boolean {
    if (sellCurrency.equals(quotedCurrency)) return true

    // 
    const wrapped = WRAPPED_NATIVE_CURRENCIES[sellCurrency.chainId as SupportedChainId]

    return Boolean(wrapped) && getIsNativeToken(sellCurrency) && quotedCurrency.equals(wrapped)
  }
  ```
  ```ts

**Review by limitofzero — APPROVED** [review]
just one nitpick, otherwise looks good, thank you!

File: apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/index.tsx (hunk 1 above)
Comment (elena-zh) [reply apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/index.tsx:124]: Good call, done in 58ce195a4.

Added `isQuotedInSameAsset` exactly as suggested, and since a mismatch is genuinely "can't tell" rather than "not enough", `getIsBalanceEnough` now returns `boolean | null`: `null` for no sell currency/amount or an unexpected currency pairing. `getSwapConfirmDisabledState` maps `null` to `{ disableConfirm: true, isInsufficientBalance: false }`, so the button is disabled without mislabelling it "Insufficient X balance".

I left out the `areBalancesReady` early-out for now — it isn't in the code today (unloaded balances read as `0n`), so it's a separate behaviour change; happy to add it here or in a follow-up if you'd prefer.

Covered by three new unit tests: no input amount, an unrelated quote currency, and a non-native sell token quoted in the wrapped native — all expect `null`.

File: apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/index.tsx
Comment (azebuado) [inline apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/index.tsx:1]: <details>
<summary>⚠️ <strong>AI Review</strong> (Claude Opus 5, worked 8m): the fix is correct, but it re-derives a balance the modal already has, via a key that is identical on every chain</summary>

I traced the fix end to end and it does work: `getCurrencyAddress` returns the native sentinel for a native currency (`libs/common-utils/src/getCurrencyAddress.ts:7-11`), and `NativeTokenBalanceUpdater` writes the wallet's native balance into `balancesAtom` under that same key (`libs/balances-and-allowances/src/updaters/NativeTokenBalanceUpdater.tsx:24-30`) — the multicall path deliberately skips native (`usePersistBalancesViaWebCalls.ts:131`), so that updater is the only thing that makes this lookup resolve. Both findings below are about how the balance is obtained, not about whether the comparison is right.

**Finding:** [NON-BLOCKING] `SwapConfirmModal` already receives the resolved input balance — the map lookup does not need to exist here

- Location: `SwapConfirmModal.utils.ts:34`, `SwapConfirmModal/index.tsx:119-131`
- The modal is already handed `inputCurrencyInfo.balance` (`CurrencyPreviewInfo.balance`), fed from `useSwapDerivedState().inputCurrencyBalance`, which is `useCurrencyAmountBalanceCombined(inputCurrency)` (`useBuildTradeDerivedState.ts:59`) — the correct native balance, already resolved as a `CurrencyAmount`. The PR instead re-derives it from the raw map with `useTokensBalancesCombined()` + `getAddressKey(getCurrencyAddress(...))`.
- That re-derivation step is exactly where #8128 went wrong. Keeping it means the bucket can be mis-picked again; taking the prop cannot.
- There is also already a shared hook for this question: `modules/combinedBalances/hooks/useHasEnoughBalanceForAmount.ts:19-31` does the identical lookup and comparison, and `TwapConfirmModal/index.tsx:63` consumes it. Root `AGENTS.md` lists "MUST prefer existing shared utilities/hooks over creating near-duplicates" under Non-Negotiables.
- Net effect: swap confirm, TWAP confirm (`useHasEnoughBalanceForAmount`, no slippage), limit confirm (`useIsZeroBalance`) and the trade form each answer "can they afford this" differently. #8128's slippage rule now lives in exactly one of the four.

**Suggested fix**
- Either pass `inputCurrencyInfo.balance` into `getIsBalanceEnough` and drop the `balances` map argument, or move the helper into `modules/combinedBalances` next to `useHasEnoughBalanceForAmount` and route both through it, with TWAP passing `maximumSellAmount: null`.

**Finding:** [QUESTION] Should the lookup keep the chain guard the equivalent hook has?

- Location: `SwapConfirmModal.utils.ts:34`, `index.tsx:121`
- `useTokensBalancesCombined()` returns `{ values, chainId }` and the call site destructures only `values`. `useCurrencyAmountBalanceCombined.ts:19` reading the same state refuses to answer when `token.chainId !== chainId`.
- This matters more after the change than before it. The wrapped address the lookup used to hit is chain-specific, so a wrong-chain read missed and fell through to `0n` — fail closed. `NATIVE_CURRENCIES[*].address` is `0xEeee…EEeE` on all eleven EVM chains, so the same read now returns another chain's native balance: a plausible wrong number that can fail *open*.
- I could not construct a repro — the modal is short-lived and `isQuoteRefreshing` short-circuits before the balance is consulted — so this is a question about the helper's contract, not a claim of a live bug.

**Suggested fix**
- Read `chainId` from the same hook and return `false` when `sellCurrency.chainId !== chainId`, matching `useCurrencyAmountBalanceCombined`. Falls out for free if the balance comes from `inputCurrencyInfo.balance` instead.

**Existing findings to handle**
- [Confirm gate not exercised in a browser](https://github.com/cowprotocol/cowswap/pull/8156#pullrequestreview-5210294763) — still open and still material; the PR body confirms the change has not been run in a browser. ER-vs-AR for a native sell on Gnosis with zero/low WXDAI, plus one wrapped-token sell as a control, would close it.

<details>
<summary>Review scope and related context</summary>

Already covered, not repeated here:
- Untested call site, and the raw cross-currency comparison — both raised in the earlier review and addressed in f2abeecd5. I re-checked both: the currency choice and the `?? inputAmount` fallback now sit inside the helper, and the comparison normalises by each side's scale.

Checked and found correct, not raised as findings:
- The scaling is right: `amountToCover.quotient * 10**sellCurrency.decimals <= balance * amountToCover.decimalScale` reduces to a plain value comparison on both sides.
- Native and wrapped decimals come from two independent SDK maps rather than being derived from one another, so normalising instead of asserting equality is a reasonable call.
- `hasCurrentCurrency` is evaluated before `isBalanceEnough` in `getSwapConfirmDisabledState`, so the helper returning `false` for a missing input amount cannot surface as a spurious "Insufficient balance".
- Missing balance still collapses to `0n`, the same as before #8156 — no change in the balances-not-yet-loaded window.
- CI is fully green on f2abeecd5, smoke included.

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

```text
Verify these findings against current code. Fix only if still valid, keep the change minimal, and validate with the targeted tests.

Context (apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/):

1. SwapConfirmModal.utils.ts:34 looks the balance up from the raw combined-balances map via
   getAddressKey(getCurrencyAddress(sellCurrency)). The modal already receives the resolved balance as
   inputCurrencyInfo.balance (CurrencyPreviewInfo.balance <- useSwapDerivedState().inputCurrencyBalance <-
   useCurrencyAmountBalanceCombined, useBuildTradeDerivedState.ts:59). A near-identical shared hook already
   exists at modules/combinedBalances/hooks/useHasEnoughBalanceForAmount.ts:19-31 and is used by
   modules/twap/containers/TwapConfirmModal/index.tsx:63.
   Expected fix: either pass inputCurrencyInfo.balance into getIsBalanceEnough and drop the balances argument,
   or move getIsBalanceEnough into modules/combinedBalances and have useHasEnoughBalanceForAmount delegate to
   it with maximumSellAmount: null. Keep the existing 9 unit tests passing.

2. The same lookup has no chain guard. useTokensBalancesCombined() also returns chainId, and
   useCurrencyAmountBalanceCombined.ts:19 returns undefined when token.chainId !== chainId. Every EVM chain's
   native address is the same 0xEeee...EEeE sentinel, so a stale/other-chain balances snapshot returns another
   chain's native balance instead of undefined.
   Expected fix: return false when sellCurrency.chainId !== the balances state chainId, or resolve finding 1
   in a way that inherits the guard. Add a test for the mismatched-chain case.

Validate with:
pnpm exec nx test cowswap-frontend --testPathPatterns=SwapConfirmModal.utils --runInBand
```

</details>

*Generated using the `pr-review` skill from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills).*

</details>

File: apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/SwapConfirmModal.utils.ts [hunk 2]
```diff
@@ -14,6 +26,23 @@ export interface SwapConfirmDisabledState {
...
+  balances,
+}: GetIsBalanceEnoughParams): boolean | null {
+  const sellCurrency = inputAmount?.currency
+  const amountToCover = maximumSellAmount ?? inputAmount
+
+  if (!sellCurrency || !amountToCover) return null
+  if (!isQuotedInSameAsset(sellCurrency, amountToCover.currency)) return null
+
+  const balance = balances[getAddressKey(getCurrencyAddress(sellCurrency))] ?? 0n
+  const sellCurrencyScale = 10n ** BigInt(sellCurrency.decimals)
```
Comment (azebuado) [inline apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/SwapConfirmModal.utils.ts]: This feels like a repetition. Can this be reused from somewhere else, maybe?

Comment (azebuado) [reply apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/SwapConfirmModal.utils.ts]: Checking it with our junior dev, there's this:

Image

Although, lines 41 and 42 won't be needed with point 1 from the AI comment I posted.
Line 43 could still benefit from the existing helper, though.

File: apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/index.tsx
Comment (elena-zh) [reply apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/index.tsx:1]: Both findings taken, in ae48baa75 — the first one resolves the second for free.

`getIsBalanceEnough` no longer touches the balances map. It takes `balance: Nullish<CurrencyAmount<Currency>>` and the modal passes `inputCurrencyInfo.balance`, i.e. the already-resolved `useCurrencyAmountBalanceCombined(inputCurrency)`. So the bucket can no longer be mis-picked — there is no bucket to pick — and the `chainId` guard on line 19 of that hook now applies here too, which closes the fail-open concern about `0xEeee…EEeE` being identical on every chain. `useTokensBalancesCombined` is gone from the modal.

A resolved-but-undefined balance (not loaded, or wrong chain) now returns `null` rather than collapsing to `0n`, so that window disables Confirm without claiming "Insufficient balance".

I did not move the helper into `modules/combinedBalances` or reroute `useHasEnoughBalanceForAmount` through it — that touches the TWAP confirm gate, which is beyond this fix. Happy to open a follow-up for the four-call-sites point; it is the more interesting half of the finding.

One trade-off worth naming: the original #8128 regression is no longer expressible as a unit test of this helper, because choosing the balance is no longer the helper's job. It is now structural — the modal is handed the input currency's balance and has nothing else to reach for.

File: apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/SwapConfirmModal.utils.ts (hunk 2 above)
Comment (elena-zh) [reply apps/cowswap-frontend/src/modules/swap/containers/SwapConfirmModal/SwapConfirmModal.utils.ts]: Your junior dev was right on all three rows. Done in ae48baa75.

Lines 41–42 are gone, as you predicted — with the resolved balance passed in there is no map lookup and no `10n ** BigInt(decimals)` to hand-roll. Line 43 now goes through the existing helpers:

```ts
return FractionUtils.lte(
  FractionUtils.fractionLikeToFraction(amountToCover),
  FractionUtils.fractionLikeToFraction(balance),
)
```

`fractionLikeToFraction` gives each side as units (`quotient / decimalScale`), so the decimals normalisation that used to be spelled out is now just what the helper does, and `lte` covers the "equalTo or lessThan" pair.

All 22 unit tests still pass, including the different-decimals case that guards the normalisation.

**Review by azebuado — APPROVED** [review]
